filters/auth/grantcookie.go

```diff
@@ -0,0 +1,126 @@
+package auth
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/zalando/skipper/secrets"
+	"golang.org/x/oauth2"
+)
+
+type cookie struct {
+	AccessToken  string    `json:"access_token"`
+	RefreshToken string    `json:"refresh_token"`
+	Expiry       time.Time `json:"expiry,omitempty"`
+	RefreshAfter time.Time `json:"refresh_after,omitempty"`
+}
+
+func refreshAfter(expiry time.Time) time.Time {
+	now := time.Now()
+	d := expiry.Sub(now)
+	if d <= 0 {
+		return now
+	}
+
+	d /= 10
+	if d < time.Minute {
+		d = time.Minute
+	}
+
+	return now.Add(d)
+}
+
+func decodeCookie(cookieHeader string, config OAuthConfig) (c *cookie, err error) {
+	var eb []byte
+	if eb, err = base64.StdEncoding.DecodeString(cookieHeader); err != nil {
+		return
+	}
+
+	var encryption secrets.Encryption
+	if encryption, err = config.Secrets.GetEncrypter(secretsRefreshInternal, config.SecretFile); err != nil {
+		return
+	}
+
+	var b []byte
+	if b, err = encryption.Decrypt(eb); err != nil {
+		return
+	}
+
+	err = json.Unmarshal(b, &c)
+	return
+}
+
+func (c cookie) isAccessTokenExpired() bool {
+	now := time.Now()
+	return now.After(c.Expiry)
+}
+
+// GetCookie extracts the OAuth Grant token cookie from a HTTP request.
+// The function supports multiple cookies with the same name and returns
+// the best match (the one that decodes properly).
+// The client may send multiple cookies if a parent domain has set a
+// cookie of the same name.
+func getCookie(request *http.Request, config OAuthConfig) (c *cookie, err error) {
+	for _, c := range request.Cookies() {
+		if c.Name == config.TokenCookieName {
+			cookie, _ := decodeCookie(c.Value, config)
+			if cookie != nil {
+				return cookie, nil
+			}
+		}
+	}
+	return nil, http.ErrNoCookie
+}
+
+// Drops the grant token cookie from the request
+func dropCookie(request *http.Request, config OAuthConfig) {
+	cookies := request.Header.Get("Cookie")
+	cookies = config.TokenCookieRegexp.ReplaceAllString(cookies, "")
+	request.Header.Set("Cookie", cookies)
+}
```

Comment:
Here I feel a bit of mismatch. getCookie tries best effort to find one and dropCookie removes all with the same name.
Should it be rather something like
```go
// extractCookie removes and returns the OAuth Grant token cookie from a HTTP request.
// The function supports multiple cookies with the same name and returns
// the best match (the one that decodes properly).
// The client may send multiple cookies if a parent domain has set a
// cookie of the same name.
func extractCookie(request *http.Request, config OAuthConfig) (cookie *cookie, err error) {
	old := request.Cookies()
	new := make([]*http.Cookie, 0, len(old))

	for i, c := range old {
		if c.Name == config.TokenCookieName {
			cookie, _ = decodeCookie(c.Value, config)
			if cookie != nil {
				new = append(new, old[i+1:]...)
				break
			}
		}
		new = append(new, c)
	}

	if cookie != nil {
		request.Header.Del("Cookie")
		for _, c := range new {
			request.AddCookie(c)
		}
		return cookie, nil
	}
	return nil, http.ErrNoCookie
}
```

Reply:
Agreed. Your approach is better. Fixed.